lightly/api/api_workflow_datasources.py

```diff
@@ -805,3 +873,119 @@ def list_datasource_permissions(
         return self._datasources_api.verify_datasource_by_dataset_id(
             dataset_id=self.dataset_id,
         ).to_dict()
+
+    def _download_raw_files(
+        self,
+        download_function: Union[
+            "DatasourcesApi.get_list_of_raw_samples_from_datasource_by_dataset_id",
+            "DatasourcesApi.get_list_of_raw_samples_predictions_from_datasource_by_dataset_id",
+            "DatasourcesApi.get_list_of_raw_samples_metadata_from_datasource_by_dataset_id",
+        ],
+        from_: int = 0,
+        to: Optional[int] = None,
+        relevant_filenames_file_name: Optional[str] = None,
+        use_redirected_read_url: bool = False,
+        progress_bar: Optional[tqdm.tqdm] = None,
+        **kwargs,
+    ) -> List[Tuple[str, str]]:
+        return list(
+            self._download_raw_files_iter(
+                download_function=download_function,
+                from_=from_,
+                to=to,
+                relevant_filenames_file_name=relevant_filenames_file_name,
+                use_redirected_read_url=use_redirected_read_url,
+                progress_bar=progress_bar,
+                **kwargs,
+            )
+        )
+
+    def _download_raw_files_iter(
+        self,
+        download_function: Union[
+            "DatasourcesApi.get_list_of_raw_samples_from_datasource_by_dataset_id",
+            "DatasourcesApi.get_list_of_raw_samples_predictions_from_datasource_by_dataset_id",
+            "DatasourcesApi.get_list_of_raw_samples_metadata_from_datasource_by_dataset_id",
+        ],
+        from_: int = 0,
+        to: Optional[int] = None,
+        relevant_filenames_file_name: Optional[str] = None,
+        use_redirected_read_url: bool = False,
+        progress_bar: Optional[tqdm.tqdm] = None,
+        **kwargs,
+    ) -> Iterator[Tuple[str, str]]:
+        if to is None:
+            to = int(time.time())
+        relevant_filenames_kwargs = (
+            {"relevant_filenames_file_name": relevant_filenames_file_name}
+            if relevant_filenames_file_name
+            else dict()
+        )
+
+        listed_filenames = set()
+
+        def get_samples(
+            response: DatasourceRawSamplesData,
+        ) -> Iterator[Tuple[str, str]]:
+            for sample in response.data:
+                if _sample_unseen_and_valid(
+                    sample=sample,
+                    relevant_filenames_file_name=relevant_filenames_file_name,
+                    listed_filenames=listed_filenames,
+                ):
+                    listed_filenames.add(sample.file_name)
+                    yield sample.file_name, sample.read_url
+                if progress_bar is not None:
```

Comment:
I left the progress bar update outside of the `if _sample_unseen_and_valid(...)` check to keep the same behavior as before. IMO it would also make sense to put it inside the `if` statement.

Reply:
I think it should be outside - we iterate over all elements of `response.data`, not just those "valid".